File: core/amber/src/main/python/core/architecture/handlers/scheduler_time_slot_event_handler.py
from core.architecture.handlers.handler_base import Handler
from core.architecture.managers.context import Context

from proto.edu.uci.ics.amber.engine.architecture.worker import SchedulerTimeSlotEventV2


class SchedulerTimeSlotEventHandler(Handler):
    cmd = SchedulerTimeSlotEventV2

    def __call__(
        self, context: Context, command: SchedulerTimeSlotEventV2, *args, **kwargs
    ):
        context.dp._scheduler_time_slot_event(command.time_slot_expired)

File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/promisehandlers/SchedulerTimeSlotEventHandler.scala
package edu.uci.ics.amber.engine.architecture.worker.promisehandlers

import edu.uci.ics.amber.engine.architecture.worker.{
  DataProcessorRPCHandlerInitializer,
  SchedulerTimeSlotExpiredPause
}
import edu.uci.ics.amber.engine.architecture.worker.promisehandlers.SchedulerTimeSlotEventHandler.SchedulerTimeSlotEvent
import edu.uci.ics.amber.engine.common.rpc.AsyncRPCServer.ControlCommand

object SchedulerTimeSlotEventHandler {
  final case class SchedulerTimeSlotEvent(timeSlotExpired: Boolean) extends ControlCommand[Unit]
}

/** Time slot start/expiration message
  *
  * possible sender: controller(by scheduler)
  */
trait SchedulerTimeSlotEventHandler {
  this: DataProcessorRPCHandlerInitializer =>

  registerHandler { (msg: SchedulerTimeSlotEvent, _) =>
    if (msg.timeSlotExpired) {
      dp.pauseManager.pause(SchedulerTimeSlotExpiredPause)
    } else {
      dp.pauseManager.resume(SchedulerTimeSlotExpiredPause)
    }
  }

}
